### PR TITLE
ssh: Parse ssh config file before connecting

### PR DESCRIPTION
**Pull request by Gundersanne**
If the (local or global) config file contains a `ProxyCommand`, the fd that cockpit interacts with is not bidirectional, and results in cockpit being broken. So this change is blocked on some api consistency on the libssh side:

- [x] https://bugs.libssh.org/T115
- [x] get libssh 0.8.5 onto fedora-testing: #10491
- [x] get libssh 0.8.5 onto fedora-29: #10520
- [x] get libssh 0.8.5 onto fedora-28: #10540
- [x] get libssh 0.8.5 onto fedora-i386: #10536
- [ ] ~get libssh 0.8.5 onto rhel-x: https://bugzilla.redhat.com/show_bug.cgi?id=1640812~ (not a blocker)
- [ ] ~get libssh 0.8.5 into Debian: https://bugs.debian.org/913242~ (not a blocker)

### CODE REVIEW

File: pkg/lib/machine-dialogs.js [hunk 1]
```diff
@@ -569,20 +569,15 @@
...
                 fp = error_options["host-fingerprint"];
-                if (error_options["invalid-hostkey-file"])
-                    path = error_options["invalid-hostkey-file"];
             }
 
             dialog.render({
                 'context_title' : dialog.context_title,
-                'path' : path,
                 'key' : fp,
-                'key_host' : key ? key.split(' ')[0] : null,
```
Comment (martinpitt) [inline pkg/lib/machine-dialogs.js]: The commit message should document the rationale for this (missing API in libssh to retrieve the file location). Could also be a separate commit for clarity.

File: src/ssh/cockpitsshoptions.c [hunk 2]
```diff
@@ -100,8 +100,7 @@ cockpit_ssh_options_from_env (gchar **env)
 {
 
   CockpitSshOptions *options = g_new0 (CockpitSshOptions, 1);
-  options->knownhosts_file = get_environment_val (env, "COCKPIT_SSH_KNOWN_HOSTS_FILE",
-                                                  default_knownhosts);
+  options->knownhosts_file = get_environment_val (env, "COCKPIT_SSH_KNOWN_HOSTS_FILE", NULL);
```
Comment (martinpitt) [inline src/ssh/cockpitsshoptions.c]: This could also be a separate commit, and even land now, right? The libssh builtin default for NULL should be exactly what we currently define as `default_knownhosts`. I think this constant and `cockpit_get_default_knownhosts()` should also be dropped, AFAICS we don't use it from anywhere.

File: src/ssh/cockpitsshrelay.c [hunk 3]
```diff
@@ -556,11 +556,11 @@ session_has_known_host_in_file (const gchar *file,
                                 const guint port)
 {
 #if !HAVE_DECL_SSH_SESSION_HAS_KNOWN_HOSTS_ENTRY
-  shim_set_knownhosts_file(file);
+  shim_set_knownhosts_file (file ? file : data->user_known_hosts);
```
Comment (martinpitt) [inline src/ssh/cockpitsshrelay.c]: The tiniest nitpick, but `value ?: fallback` is a bit more succinct. (SCNR)

File: src/ssh/cockpitsshrelay.c [hunk 4]
```diff
@@ -556,11 +556,11 @@ session_has_known_host_in_file (const gchar *file,
...
 #endif
   /* HACK: https://bugs.libssh.org/T108 */
   if (!file)
-      g_warn_if_fail (ssh_options_set (data->session, SSH_OPTIONS_SSH_DIR, NULL) == SSH_OK);
+    g_warn_if_fail (ssh_options_set (data->session, SSH_OPTIONS_SSH_DIR, NULL) == SSH_OK);
   g_warn_if_fail (ssh_options_set (data->session, SSH_OPTIONS_KNOWNHOSTS, file) == SSH_OK);
```
Comment (martinpitt) [inline src/ssh/cockpitsshrelay.c]: BTW, shouldn't that be `SSH_OPTIONS_GLOBAL_KNOWNHOSTS`?

File: src/ssh/test-sshbridge.c [hunk 5]
```diff
@@ -284,6 +286,29 @@ setup (TestCase *tc,
...
 
+  if (fixture && fixture->test_home_ssh_config)
+    {
+      gchar *content;
+
+      host = g_strdup ("some_host");
+      argv[1] = host;
+
+      tc->home_ssh_dir = g_build_filename (tc->home_dir, ".ssh", NULL);
+      g_assert (tc->home_ssh_dir != NULL);
```
Comment (martinpitt) [inline src/ssh/test-sshbridge.c]: This duplicates stuff from above, let's just move it further up and do it unconditionally.

File: pkg/lib/machine-dialogs.js (hunk 1 above)
Comment (Gundersanne) [reply pkg/lib/machine-dialogs.js]: That make sense, I'll separate it.

File: src/ssh/cockpitsshoptions.c (hunk 2 above)
Comment (Gundersanne) [reply src/ssh/cockpitsshoptions.c]: I'm not sure if it can land right now, first libssh should check the global known_hosts file by default. 

Only then can we stop checking the global known_hosts file explicitly post 0.8 (and stop using `ssh_options->knownhosts_file` for both global known_hosts file and a value set through the env).

File: src/ssh/cockpitsshrelay.c (hunk 3 above)
Comment (Gundersanne) [reply src/ssh/cockpitsshrelay.c]: Haha cool, I didn't even knew that syntax existed :D

File: src/ssh/test-sshbridge.c (hunk 5 above)
Comment (Gundersanne) [reply src/ssh/test-sshbridge.c]: Oki will do

File: src/ssh/cockpitsshrelay.c (hunk 4 above)
Comment (Gundersanne) [reply src/ssh/cockpitsshrelay.c]: Oh yea you're right, after T107 we can use that option.

**Comment by Gundersanne** [comment]
Libssh just did a release, so we should be ready to go after the next refreshes.

**Comment by martinpitt** [comment]
I clarified the "image refreshes" TODO items with a list of particular images, with pointers to PRs. I think we should get 0.8.5 at least on these, we can then handle the others through known issues.

**Comment by Gundersanne** [comment]
@pitti So should we avoid parsing the config file on <0.8.5? Because before that a proxycommand will break cockpit.

**Comment by martinpitt** [comment]
@Gundersanne : Yes, I think that's a good compromise. Then it at least doesn't break on older releases, and this would get unblocked. I think we now just need to rebuild the fedora-{28,i386} images, then you have enough fodder to play with :-)

File: src/ssh/test-sshbridge.c [hunk 6]
```diff
@@ -284,6 +287,28 @@ setup (TestCase *tc,
       g_free (content);
     }
 
+#if HAVE_SAFE_PROXY_COMMANDS
+  if (fixture && fixture->test_home_ssh_config)
+    {
+      gchar *content;
+
+      host = g_strdup ("some_host");
```
Comment (martinpitt) [inline src/ssh/test-sshbridge.c]: This leaks the previous value of `host`, so it needs to be freed first. But moreover, I think the original value should be plugged into `content`, so I recommend to build `content` string first with the original `host` value, then free it, and set it to `some_host`.

Comment (martinpitt) [reply src/ssh/test-sshbridge.c]: ... curious that this doesn't get picked up by our `check-memory` semaphore test run. Am I missing something here?

File: test/verify/check-multi-machine [hunk 7]
```diff
@@ -192,6 +192,30 @@ class TestMultiMachineAdd(MachineCase):
                                     ".*: bridge failed: .*",
                                     ".*: bridge program failed: Child process exited with code .*")
 
+    @skipImage("Newer version (>= 0.8.5) of libssh required", "debian-stable", "debian-testing",
```
Comment (martinpitt) [inline test/verify/check-multi-machine]: debian-testing does have 0.8.5, that should be fine?

Comment (martinpitt) [reply test/verify/check-multi-machine]: ignore me, it doesn't.

**Comment by martinpitt** [comment]
Please run this manually and fix whatever it's complaining about:

   FAIL: tools/test-static-code 2 pyflakes test

**Comment by martinpitt** [comment]
Tests were green before (except for pyflakes), and all current failures are unrelated to c-ssh.
